File: Data Structures & Algorithms/binary-tree-right-side-view/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        Deque<TreeNode> dq = new LinkedList<>();
        Queue<TreeNode> q = new LinkedList<>();

        dq.offerLast(root);

        while (!dq.isEmpty()) {
            res.add(dq.peekLast().val);
            while (!dq.isEmpty()) {
                q.offer(dq.poll());
            }
            while (!q.isEmpty()) {
                TreeNode node = q.poll();
                if (node.left != null) {
                    dq.offerLast(node.left);
                }
                if (node.right != null) {
                    dq.offerLast(node.right);
                }
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun levelOrder(root: TreeNode?): List<List<Int>> {
        val result = mutableListOf<List<Int>>()
        val q = ArrayDeque<TreeNode>()

        if (root == null) {
            return emptyList()
        }

        q.add(root)

        while (!q.isEmpty()) {
            val qSize = q.size
            val valList = mutableListOf<Int>()
            for (i in 0 until qSize) {
                val node = q.removeFirst()
                valList.add(node.`val`)

                if (node.left != null) {
                    q.add(node.left)
                }
                if (node.right != null) {
                    q.add(node.right)
                }
            }
            result.add(valList)
        }
        return result
    }
}
